### DISCUSSION TITLE
How to order the points that returned by extract_feature_edges?

### DISCUSSION THREAD

**Discussion by SivHMA** [pyvista/pyvista#3211] [Q&A] [answered]
I extract the border edges by extract_feature_edges, and I find that the returned points are not directionally ordered. Or I made a mistake?
`edges = mesh.extract_feature_edges(boundary_edges=True, feature_edges=False, manifold_edges=False)
`

![image](https://user-images.githubusercontent.com/29498630/186626349-a0127015-f968-46d7-ae9a-a0947e02f9ee.png)

`spline = pv.Spline(edges.points, 1000)`
![image](https://user-images.githubusercontent.com/29498630/186626623-d4b6f4de-4be1-4f80-9ef2-af0ab3433846.png)

Is there any way to sort the points?

**Comment by darikg** [comment, marked as answer]
I don't know of an easy pyvista/vtk solution. libigl has  [boundary_loop](https://libigl.github.io/libigl-python-bindings/igl_docs/#boundary_loop) and [edges_to_path](https://libigl.github.io/libigl-python-bindings/igl_docs/#boundary_loop) which construct ordered vertex indices.

Here's my untested no-dependency solution:

```python
edge_data = mesh.extract_feature_edges(boundary_edges=True, feature_edges=False, manifold_edges=False)
edges = edge_data.lines.reshape(-1, 3)[:, 1:]  # n_edges x 2
adj = {u: v for u, v in edges}  # Adjacency map
v = edges[0, 0]  # Starting vertex
vert_idxs = [v]
for _ in range(len(edges)):
    v = adj[v]
    vert_idxs.append(v)

points = edge_data.points[vert_idxs]
```

**Reply by akaszynski** [reply]
Here's an expanded solution that builds off of @darikg's answer:

```py
import pyvista as pv
from pyvista import examples
import numpy as np

dataset = examples.download_bunny_coarse()
mesh = dataset.clip('y', invert=True)

edge_data = mesh.extract_feature_edges(boundary_edges=True, feature_edges=False, manifold_edges=False)


def reorder_edges(edge_data):
    edges = edge_data.lines.reshape(-1, 3)[:, 1:]  # n_edges x 2

    u, v = edges.T

    adj = np.empty(u.shape, dtype=u.dtype)
    adj[u] = v

    v = edges[0, 0]  # Starting vertex
    vert_idxs = [v]
    for _ in range(len(edges)):
        v = adj[v]
        vert_idxs.append(v)

    new_lines = np.empty((edges.shape[0], 3), dtype=edges.dtype)
    new_lines[:, 0] = 2
    new_lines[:, 1] = range(edges.shape[0])
    new_lines[:, 2] = range(1, edges.shape[0] + 1)
    return pv.PolyData(edge_data.points[vert_idxs], lines=new_lines)


new_edges = reorder_edges(edge_data)

pl = pv.Plotter()
pl.add_mesh(mesh)
pl.add_mesh(new_edges, scalars=range(new_edges.n_points), render_lines_as_tubes=True, line_width=5, show_scalar_bar=False)
pl.show()
```
![tmp](https://user-images.githubusercontent.com/11981631/186717963-024c6aaf-a979-46dc-8b02-d5dc574d084a.png)

**Reply by SivHMA** [reply]
thanks!

**Reply by akaszynski** [reply]
Please note @adeak's caveats in https://github.com/pyvista/pyvista/pull/3213.

**Reply by adam-grant-hendry** [reply]
> Please note @adeak's caveats in https://github.com/pyvista/pyvista/pull/3213.

In general then, unless `libigl` is reverse-engineered into `pyvista`, a "fuller-proof" approach is as follows (combining everything here):

```python
import igl
import numpy as np
import pyvista as pv

def orient_edges(curve: pv.PolyData) -> pv.PolyData:
    edges = curve.lines.reshape(-1, 3)[:, 1:]
    path, _, _ = igl.edges_to_path(edges)

    lines = np.empty((edges.shape[0], 3), dtype=edges.dtype)
    lines[:, 0] = 2  # Assumes number of points in edge is always 2
    lines[:, 1] = range(edges.shape[0])
    lines[:, 2] = range(1, edges.shape[0] + 1)

    return pv.PolyData(curve.points[path], lines=lines)
```
